File: backend/src/main/java/br/ufpr/sept/so2/modules/iam/infrastructure/persistence/SenhaHistoricoJpaAdapter.java
package br.ufpr.sept.so2.modules.iam.infrastructure.persistence;

import br.ufpr.sept.so2.modules.iam.application.ports.SenhaHistoricoRepository;
import br.ufpr.sept.so2.shared.infrastructure.Uuids;
import org.springframework.stereotype.Component;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

@Component
public class SenhaHistoricoJpaAdapter implements SenhaHistoricoRepository {

    private final SenhaHistoricoJpaRepository jpaRepository;

    public SenhaHistoricoJpaAdapter(SenhaHistoricoJpaRepository jpaRepository) {
        this.jpaRepository = jpaRepository;
    }

    @Override
    public void append(UUID usuarioId, String senhaHash) {
        jpaRepository.save(new SenhaHistoricoJpaEntity(Uuids.v7(), usuarioId, senhaHash, OffsetDateTime.now()));
    }

    @Override
    public List<String> findLastHashes(UUID usuarioId, int limite) {
        return jpaRepository.findTop3ByUsuarioIdOrderByCreatedAtDesc(usuarioId).stream()
                .limit(limite)
                .map(SenhaHistoricoJpaEntity::getSenhaHash)
                .toList();
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/iam/infrastructure/persistence/SenhaHistoricoJpaAdapter.kt
package br.ufpr.sept.so2.modules.iam.infrastructure.persistence

import br.ufpr.sept.so2.modules.iam.application.ports.SenhaHistoricoRepository
import br.ufpr.sept.so2.shared.infrastructure.Uuids
import org.springframework.stereotype.Component
import java.time.OffsetDateTime
import java.util.UUID

@Component
class SenhaHistoricoJpaAdapter(
    private val jpaRepository: SenhaHistoricoJpaRepository,
) : SenhaHistoricoRepository {
    override fun append(usuarioId: UUID, senhaHash: String) {
        jpaRepository.save(SenhaHistoricoJpaEntity(Uuids.v7(), usuarioId, senhaHash, OffsetDateTime.now()))
    }

    override fun findLastHashes(usuarioId: UUID, limite: Int): List<String> =
        jpaRepository.findTop3ByUsuarioIdOrderByCreatedAtDesc(usuarioId)
            .asSequence()
            .take(limite)
            .map { it.senhaHash!! }
            .toList()
}
